src/lazy/text/parse_result.rs

```diff
@@ -0,0 +1,274 @@
+//         if let Some(label) = value.label {
+//             message.push_str(" while ");
+//             message.push_str(label.as_ref());
+//         }
+//         let position = Position::with_offset(value.input.offset()).with_length(value.input.len());
+//         let decoding_error = DecodingError::new(message).with_position(position);
+//         IonError::Decoding(decoding_error)
+//     }
+// }
+
+impl<'data> From<InvalidInputError<'data>> for IonParseError<'data> {
+    fn from(value: InvalidInputError<'data>) -> Self {
+        IonParseError::Invalid(value)
+    }
+}
+
+impl<'data> From<nom::Err<IonParseError<'data>>> for IonParseError<'data> {
+    fn from(value: Err<IonParseError<'data>>) -> Self {
+        match value {
+            Err::Incomplete(_) => IonParseError::Incomplete,
+            Err::Error(e) => e,
+            Err::Failure(e) => e,
+        }
+    }
+}
+
+/// Allows an `IonParseError` to be constructed from a `(&str, ErrorKind)` tuple, which is the
+/// data provided by core `nom` parsers if they do not match the input.
+impl<'data> From<(TextBufferView<'data>, ErrorKind)> for IonParseError<'data> {
+    fn from((input, error_kind): (TextBufferView<'data>, ErrorKind)) -> Self {
+        InvalidInputError::new(input)
+            .with_nom_error_kind(error_kind)
+            .into()
+    }
+}
+
+/// Allows a [nom::error::Error] to be converted into an [IonParseError] by calling `.into()`.
+impl<'data> From<NomError<TextBufferView<'data>>> for IonParseError<'data> {
+    fn from(nom_error: NomError<TextBufferView<'data>>) -> Self {
+        InvalidInputError::new(nom_error.input)
+            .with_nom_error_kind(nom_error.code)
+            .into()
+    }
+}
+
+/// Allows `IonParseError` to be used as the error type in various `nom` functions.
+impl<'data> ParseError<TextBufferView<'data>> for IonParseError<'data> {
+    fn from_error_kind(input: TextBufferView<'data>, error_kind: ErrorKind) -> Self {
+        InvalidInputError::new(input)
+            .with_nom_error_kind(error_kind)
+            .into()
+    }
+
+    fn append(input: TextBufferView<'data>, kind: ErrorKind, mut other: Self) -> Self {
+        // When an error stack is being built, this method is called to give the error
+        // type an opportunity to aggregate the errors into a collection or a more descriptive
+        // message. For now, we simply allow the most recent error to take precedence.
+        let new_error = InvalidInputError::new(input).with_nom_error_kind(kind);
+        if let IonParseError::Invalid(invalid_input_error) = &mut other {
+            invalid_input_error.backtrace.push(new_error)
+        }
+        other
+    }
+}
+
+pub(crate) trait AddContext<'data, T> {
+    fn with_context(
+        self,
+        label: impl Into<Cow<'static, str>>,
+        input: TextBufferView<'data>,
+    ) -> IonResult<(TextBufferView<'data>, T)>;
+}
```

Comment:
🗺️ I plan to add a doc comment to this trait before merging. I intentionally avoided providing an `Into<IonError>` impl for `IonParseError` because some of the nom error kinds that bubble up as `IonParseError` don't retain context. This trait is an easy way for methods that could surface without context to require more information before being passed on to the user.

For an example of its usage, see `LazyRawTextReader::next()`:
![image](https://github.com/amazon-ion/ion-rust/assets/611616/edf73781-74c6-42ea-a1dd-90978d68af7b)